sys/net/application_layer/emcute/emcute.c

```diff
@@ -0,0 +1,568 @@
+        res = syncsend(PUBACK, len, true);
+    }
+    else {
+        sock_udp_send(&sock, tbuf, len, &gateway);
+        mutex_unlock(&txlock);
+    }
+
+    return res;
+}
+
+int emcute_sub(emcute_sub_t *sub, unsigned flags)
+{
+    assert(sub && (sub->cb) && (sub->topic.name) && !(flags & ~SUB_FLAGS));
+
+    if (gateway.port == 0) {
+        return EMCUTE_NOGW;
+    }
+    if (strlen(sub->topic.name) > EMCUTE_TOPIC_MAXLEN) {
+        return EMCUTE_OVERFLOW;
+    }
+
+    mutex_lock(&txlock);
+
+    tbuf[0] = (strlen(sub->topic.name) + 5);
+    tbuf[1] = SUBSCRIBE;
+    tbuf[2] = flags;
+    set_u16(&tbuf[3], id_next);
+    waitonid = id_next++;
+    memcpy(&tbuf[5], sub->topic.name, strlen(sub->topic.name));
+
+    int res = syncsend(SUBACK, (size_t)tbuf[0], false);
+    if (res > 0) {
+        DEBUG("[emcute] sub: success, topic id is %i\n", res);
+        sub->topic.id = res;
+
+        /* check if subscription is already in the list, only insert if not*/
+        emcute_sub_t *s;
+        for (s = subs; s && (s != sub); s = s->next) {}
+        if (!s) {
+            sub->next = subs;
+            subs = sub;
+            res = EMCUTE_OK;
+        }
+    }
+
+    mutex_unlock(&txlock);
+    return res;
+}
+
+int emcute_unsub(emcute_sub_t *sub)
+{
+    assert(sub && sub->topic.name);
+
+    if (gateway.port == 0) {
+        return EMCUTE_NOGW;
+    }
+
+    mutex_lock(&txlock);
+
+    tbuf[0] = (strlen(sub->topic.name) + 5);
+    tbuf[1] = UNSUBSCRIBE;
+    tbuf[2] = 0;
+    set_u16(&tbuf[3], id_next);
+    waitonid = id_next++;
+    memcpy(&tbuf[5], sub->topic.name, strlen(sub->topic.name));
+
+    int res = syncsend(UNSUBACK, (size_t)tbuf[0], false);
+    if (res == EMCUTE_OK) {
+        if (subs == sub) {
+            subs = sub->next;
+        }
+        else {
+            emcute_sub_t *s;
+            for (s = subs; s; s = s->next) {
+                if (s->next == sub) {
+                    s->next = sub->next;
+                    break;
+                }
+            }
+        }
+    }
+
+    mutex_unlock(&txlock);
+    return res;
+}
+
+int emcute_willupd_topic(const char *topic, unsigned flags)
+{
+    assert(!(flags & ~PUB_FLAGS));
+
+    if (gateway.port == 0) {
+        return EMCUTE_NOGW;
+    }
+    if (topic && (strlen(topic) > EMCUTE_TOPIC_MAXLEN)) {
+        return EMCUTE_OVERFLOW;
+    }
+
+    mutex_lock(&txlock);
+
+    tbuf[1] = WILLTOPICUPD;
+    if (!topic) {
+        tbuf[0] = 2;
+    }
+    else {
+        tbuf[0] = (strlen(topic) + 3);
+        tbuf[2] = flags;
+        memcpy(&tbuf[3], topic, strlen(topic));
+    }
+
+    return syncsend(WILLTOPICRESP, (size_t)tbuf[0], true);
+}
+
+int emcute_willupd_msg(const void *data, size_t len)
+{
+    assert(data && (len > 0));
+
+    if (gateway.port == 0) {
+        return EMCUTE_NOGW;
+    }
+    if (len > (EMCUTE_BUFSIZE - 4)) {
+        return EMCUTE_OVERFLOW;
+    }
+
+    mutex_lock(&txlock);
+
+    int pos = set_len(tbuf, (len + 1));
+    len += (pos + 1);
```

Comment:
@haukepetersen 
This `len += (pos + 1);` looks suspicious. Coverity Scan tells us that there is a potential buffer overflow.

If you ask me, I would say that this statement should not there. And if it is intended then the if condition above needs to be changed.

Reply:
`len` comes from a local user call. This could just be solved by an appropriate `assert()`.

Reply:
Maybe I wasn't clear. My point is that this statement should not be there at all.

`len` is the length of the data. It does not make sense to increase it by 4. The three bytes for the length field are already copied into `tbuf` through `set_len`. Then `WILLMSGUPD` is copied into the buffer. What is left to do with `memcpy` is to copy the data with length `len`. Notice the `&tbuf[pos]` being the destination of `memcpy`.

Reply:
@haukepetersen any comment to this?

Reply:
See issue #15028 to discuss this further